pkg/aws/commandbuilder.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+type Service string
+
+const (
+	IAM Service = "iam"
+)
+
+type Command string
+
+const (
+	CREATE_ROLE                      Command = "create-role"
+	DELETE_ROLE                      Command = "delete-role"
+	CREATE_POLICY                    Command = "create-policy"
+	DELETE_POLICY                    Command = "delete-policy"
+	CREATE_POLICY_VERSION            Command = "create-policy-version"
+	DELETE_ROLE_POLICY               Command = "delete-role-policy"
+	ATTACH_ROLE_POLICY               Command = "attach-role-policy"
+	DETACH_ROLE_POLICY               Command = "detach-role-policy"
+	TAG_POLICY                       Command = "tag-policy"
+	TAG_ROLE                         Command = "tag-role"
+	CREATE_OPEN_ID_CONNECT_PROVIDER  Command = "create-open-id-connect-provider"
+	DELETE_OPEN_ID_CONNECT_PROVIDER  Command = "delete-open-id-connect-provider"
+	DELETE_ROLE_PERMISSIONS_BOUNDARY Command = "delete-role-permissions-boundary"
+)
+
+type Param string
+
+const (
+	TAGS                         Param = "tags"
+	ROLE_NAME                    Param = "role-name"
+	ASSUME_ROLE_POLICY_DOCUMENT  Param = "assume-role-policy-document"
+	PERMISSIONS_BOUNDARY         Param = "permissions-boundary"
+	PATH                         Param = "path"
+	POLICY_NAME                  Param = "policy-name"
+	POLICY_DOCUMENT              Param = "policy-document"
+	POLICY_ARN                   Param = "policy-arn"
+	URL                          Param = "url"
+	CLIENT_ID_LIST               Param = "client-id-list"
+	THUMBPRINT_LIST              Param = "thumbprint-list"
+	OPEN_ID_CONNECT_PROVIDER_ARN Param = "open-id-connect-provider-arn"
+	SET_AS_DEFAULT               Param = "set-as-default"
+)
+
+type ICommandBuilder interface {
+	SetService(service Service)
+	SetCommand(command Command)
+	AddParam(param Param, value interface{})
+	AddParamNoValue(param Param)
+	Build() string
+}
+
+type CommandBuilder struct {
+	service Service
+	command Command
+	params  []string
+}
+
+func (b *CommandBuilder) SetService(awsService Service) {
+	b.service = awsService
+}
+
+func (b *CommandBuilder) SetCommand(awsCommand Command) {
+	b.command = awsCommand
+}
+
+type StringMap map[string]string
+
+func (b *CommandBuilder) AddParam(awsParam Param, value interface{}) {
```

Comment:
how about splitting this to two methods, like so:
```suggestion
func (b *CommandBuilder) AddParam(awsParam Param, value string) {...
func (b *CommandBuilder) AddKVParam(awsParam Param, value map[string]string) {...
```
Or even, if i'm not mistaken and these are always used for `--tags` parameter:
```suggestion
func (b *CommandBuilder) AddParam(awsParam Param, value string) {...
func (b *CommandBuilder) AddTags(tags map[string]string) {...
```
and add a `tags []string` to the builder struct?
That will allow us to use the compile time type safety, make the methods clearer and allow conditionally adding more tags (e.g. for the `isAdmin` case)

Reply:
Coming from overloading I just wanted to have a single entry point to params, but this is ok by me as well. Will update it following this setup